lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -26,10 +26,60 @@ using namespace circt;
 using namespace circt::calyx;
 using namespace mlir;
 
+//===----------------------------------------------------------------------===//
+// ProgramOp
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verifyProgramOp(ProgramOp program) {
+  if (!program.getMainComponent())
+    return program.emitOpError("Must contain one component named "
+                               "\"main\" as the entry point.");
+  return success();
+}
+
 //===----------------------------------------------------------------------===//
 // ComponentOp
 //===----------------------------------------------------------------------===//
 
+/// Returns the type of the given component as a function type.
+static FunctionType getComponentType(Operation *component) {
+  auto typeAttr =
+      component->getAttrOfType<TypeAttr>(ComponentOp::getTypeAttrName());
+  return typeAttr.getValue().cast<FunctionType>();
+}
+
+/// Returns the component port names in the given direction.
+static ArrayAttr getComponentPortNames(Operation *component,
```

Comment:
I think if this was a `ComponentOp` rather than an opaque `Operation *`, you would have access to named helpers you could call to get the relevant attribute, rather than `getAttrOfType`.